simul/cothority/cothority.go

```diff
@@ -104,24 +113,37 @@ func main() {
 			dbg.Fatal(err)
 		}
 
-		// In case of "SingleHost" we need a new tree that contains every
-		// entity only once, whereas rootSC.Tree will have the same
-		// entity at different TreeNodes, which makes it difficult to
-		// correctly close everything.
-		closeTree := rootSC.EntityList.GenerateBinaryTree()
-		rootSC.Overlay.RegisterTree(closeTree)
-		_, err = rootSC.Overlay.StartNewNodeName("CloseAll", closeTree)
+		if rootSC.IsSingleHost() {
+			// In case of "SingleHost" we need a new tree that contains every
+			// entity only once, whereas rootSC.Tree will have the same
+			// entity at different TreeNodes, which makes it difficult to
+			// correctly close everything.
+			dbg.Lvl2("Making new root-tree for SingleHost config")
+			closeTree := rootSC.EntityList.GenerateBinaryTree()
+			rootSC.Overlay.RegisterTree(closeTree)
+			_, err = rootSC.Overlay.StartNewNodeName("CloseAll", closeTree)
+		} else {
+			_, err = rootSC.Overlay.StartNewNodeName("CloseAll", rootSC.Tree)
+		}
+		monitor.End()
 		if err != nil {
 			dbg.Fatal(err)
 		}
-	} else {
-		rootSC = scs[0]
 	}
 
-	// Wait for the first host to be closed
+	// Wait for all hosts to be closed
+	allClosed := make(chan bool, 1)
+	go func() {
+		for _, sc := range scs {
+			<-sc.Host.Closed
+			dbg.Lvl4("Host", sc.Host.Entity.Addresses, "closed")
+		}
+		allClosed <- true
+	}()
 	select {
-	case <-rootSC.Host.Closed:
-		dbg.Lvl2(hostAddress, ": first host closed - quitting")
-		monitor.End()
+	case <-allClosed:
+		dbg.Lvl2(hostAddress, ": all hosts closed")
+	case <-time.After(time.Minute * 2):
```

Comment:
Make this timeout a global var (const) or as you suggested put them into toml files (with reasonable default if there is nothing set in the toml)